✨ Prefill title and message in hook mode

adriencaccia opened the pull request:
## Description

Hi @carloscuesta, this PR adds prefilling the `title` and `message` prompts when in `hook` mode.
This allows to keep your titles and messages in your history.

![gitmoji-prefill-hook](https://user-images.githubusercontent.com/19605940/90916131-d42b2c00-e3e0-11ea-8029-c977399b12f7.gif)

adriencaccia commented:
@carloscuesta Hello, I tried to have 100% in diff coverage, however it seems that the `prompts.js` file is not counted in the coverage (it might so because it is a snapshot test).

Given that the coverage was already low for this file, would you say that the tests I have written are sufficient ?

carloscuesta commented on .vscode/settings.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "javascript.validate.enable": false,
+  "typescript.validate.enable": false,
+  "editor.formatOnSave": true
+}
```
Let's put this on the `.gitignore` file :pray:

carloscuesta commented on src/commands/commit/prompts.js:
```diff
@@ -23,40 +24,51 @@ export type Answers = {
+      validate: guard.title,
+      transformer: (input: string) => {
+        return `[${
+          (title || input).length
+        }/${TITLE_MAX_LENGTH_COUNT}]: ${input}`
+      },
+      ...(title ? { default: title } : {})
+    },
+    {
+      name: 'message',
+      message: 'Enter the commit message:',
+      validate: guard.message,
+      ...(message ? { default: message } : {})
     }
-  },
-  ...(configurationVault.getScopePrompt()
-    ? [
-        {
-          name: 'scope',
-          message: 'Enter the scope of current changes:',
-          validate: guard.scope
-        }
-      ]
-    : []),
-  {
-    name: 'title',
-    message: 'Enter the commit title',
-    validate: guard.title,
-    transformer: (input: string) => {
-      return `[${input.length}/${TITLE_MAX_LENGTH_COUNT}]: ${input}`
-    }
-  },
-  {
-    name: 'message',
-    message: 'Enter the commit message:',
-    validate: guard.message
-  }
-]
+  ]
+}
```
Things we can improve here:

- The `client` | `mode` type should be extracted into the `index` file and imported here to not repeat over and over again

```
// @flow
import inquirer from 'inquirer'

import configurationVault from '../../utils/configurationVault'
import filterGitmojis from '../../utils/filterGitmojis'
import getDefaultCommitContent from '../../utils/getDefaultCommitContent'
import { type CommitMode } from './index'
import guard from './guard'

const TITLE_MAX_LENGTH_COUNT: number = 48

inquirer.registerPrompt('autocomplete', require('inquirer-autocomplete-prompt'))

export type Gitmoji = {
  code: string,
  description: string,
  emoji: string,
  name: string
}

export type Answers = {
  gitmoji: string,
  scope?: string,
  title: string,
  message: string
}

export default (gitmojis: Array<Gitmoji>, mode: CommitMode): Array<Object> => [
  {
    name: 'gitmoji',
    message: 'Choose a gitmoji:',
    type: 'autocomplete',
    source: (answersSoFor: any, input: string) => {
      return Promise.resolve(
        filterGitmojis(input, gitmojis).map((gitmoji) => ({
          name: `${gitmoji.emoji}  - ${gitmoji.description}`,
          value: gitmoji[configurationVault.getEmojiFormat()]
        }))
      )
    }
  },
  ...(configurationVault.getScopePrompt()
    ? [
        {
          name: 'scope',
          message: 'Enter the scope of current changes:',
          validate: guard.scope
        }
      ]
    : []),
  {
    name: 'title',
    message: 'Enter the commit title',
    validate: guard.title,
    transformer: (input: string) => {
      return `[${
        (title || input).length
      }/${TITLE_MAX_LENGTH_COUNT}]: ${input}`
    },
    default: mode === 'hook'
      ? getDefaultCommitContent().title
      : undefined
  },
  {
    name: 'message',
    message: 'Enter the commit message:',
    validate: guard.message,
    default: mode === 'hook'
      ? getDefaultCommitContent().message
      : undefined
  }
]
```

adriencaccia replied:
Ok for the type !
For the `default` properties definition on the other hand, I will have to keep the syntax where I initialize `title` and `message` before returning the array.
Indeed, with a syntax like the one you suggested, when running a simple `git commit` without messages, the user will be shown the following, resulting of `default: undefined`
![image](https://user-images.githubusercontent.com/19605940/93653616-f6cd5680-fa19-11ea-814b-3f33461335d6.png)

Moreover, with this syntax, the `getDefaultCommitContent` will be called only once

carloscuesta replied:
Yes! Makes sense to keep then the variables on the function 😊

carloscuesta commented on src/utils/getDefaultCommitContent.js:
```diff
@@ -0,0 +1,23 @@
+
+const getDefaultCommitContent = () => {
+  try {
+    const commitFilePath: string = process.argv[3]
```
I think we need to clarify or move those magic numbers into constants that explain why we use those values

carloscuesta commented on src/utils/getDefaultCommitContent.js:
```diff
@@ -0,0 +1,23 @@
+      .toString()
+      .split('\n')
+    return {
+      title: commitFileContent.length ? commitFileContent[0] : '',
+      message: commitFileContent.length >= 3 ? commitFileContent[2] : ''
```
I think we need to clarify or move those magic numbers into constants that explain why we use those values

carloscuesta commented on src/utils/getDefaultCommitContent.js:
```diff
@@ -0,0 +1,23 @@
+      title: commitFileContent.length ? commitFileContent[0] : '',
+      message: commitFileContent.length >= 3 ? commitFileContent[2] : ''
+    }
+  } catch (e) {
```
We're catching the error and not doing anything ? 🤔

carloscuesta commented on test/utils/getDefaultCommitContent.spec.js:
```diff
@@ -0,0 +1,35 @@
+import getDefaultCommitContent from '../../src/utils/getDefaultCommitContent'
+
+describe('getDefaultCommitContent', () => {
+  it('should retrieve title and message when file exists', () => {
```
I would move those it on different describes:

describe(when file exists)
  it(should retreive title and message)

describe(when file does not exists)
  it(should retrieve ...)

carloscuesta reviewed (commented):
👀  Some questions and comments!

Thanks for sending a PR ❤️

adriencaccia commented:
I implemented the changes you requested, and replaced magic numbers in `getDefaultCommitContent.js` with a comment and some constants to better explain their roles.

carloscuesta commented on test/utils/getDefaultCommitContent.spec.js:
```diff
@@ -0,0 +1,34 @@
+describe('getDefaultCommitContent', () => {
+  describe('when file exists', () => {
+    beforeAll(() => {
+      process.argv[3] = 'commit'
```
It's not dangerous to overwrite the `process.argv` inside of tests? 🤔

adriencaccia replied:
Since we mock the `fs` functions that use the value of `process.argv[3]`, I think we can just not bother ovrewriting it at all

carloscuesta commented on test/utils/getDefaultCommitContent.spec.js:
```diff
@@ -0,0 +1,34 @@
+
+  describe('when file does not exist', () => {
+    beforeAll(() => {
+      process.argv[3] = 'commit'
```
It's not dangerous to overwrite the `process.argv` inside of tests? 🤔

carloscuesta commented on src/commands/commit/prompts.js:
```diff
@@ -23,40 +25,50 @@ export type Answers = {
-    transformer: (input: string) => {
-      return `[${input.length}/${TITLE_MAX_LENGTH_COUNT}]: ${input}`
+export default (gitmojis: Array<Gitmoji>, mode: CommitMode): Array<Object> => {
+  const { title, message } = mode === 'hook'
+    ? getDefaultCommitContent()
+    : { title: undefined, message: undefined }
```
I have a question right here, If we are returning `{ title: undefined, message: undefined }` in here, what's the point then of returning it also inside of the `getDefaultCommitContent` function ?

I mean can we find another way to remove the default values for title and message prompts where the mode it's not hook ?

adriencaccia replied:
I added the `mode` parameter to `getDefaultCommitContent` util so that we can return null values directly in the function when `mode === 'client'`

carloscuesta commented on src/utils/getDefaultCommitContent.js:
```diff
@@ -0,0 +1,36 @@
+  */
+  const commitFilePath: string = process.argv[COMMIT_FILE_PATH_INDEX]
+
+  if (fs.existsSync(commitFilePath) === false) {
+    return {
+      title: undefined,
+      message: undefined
+    }
+  }
```
This is for when the hook command it's not autocompleted right ? Can we use for example `null` instead of `undefined` in here to remove the default values for inquirer prompt ?

I think we should use `null` instead of `undefined`

adriencaccia replied:
Done

carloscuesta commented on src/utils/getDefaultCommitContent.js:
```diff
@@ -0,0 +1,36 @@
+const COMMIT_TITLE_LINE_INDEX = 0
+const COMMIT_MESSAGE_LINE_INDEX = 2
+
+const getDefaultCommitContent = () => {
```
It would be cool to type the return value of this function

```suggestion
const getDefaultCommitContent = (): { title: type, message: type } => {
```

adriencaccia replied:
Done

carloscuesta commented on src/utils/getDefaultCommitContent.js:
```diff
@@ -0,0 +1,36 @@
+  */
+  const commitFilePath: string = process.argv[COMMIT_FILE_PATH_INDEX]
+
+  if (fs.existsSync(commitFilePath) === false) {
```
```suggestion
  if (!fs.existsSync(commitFilePath)) {
```

adriencaccia replied:
Done

carloscuesta reviewed (commented):
👏🏼

carloscuesta reviewed (approved):
I think we're ready to merge this 🚀 

Thanks for the work you've put on this PR @adriencaccia ❤️

Larandar commented:
Thank you ❤️
